Simplenote/src/test/java/com/automattic/simplenote/viewmodels/NoteEditorViewModelTest.kt

```diff
@@ -0,0 +1,87 @@
+package com.automattic.simplenote.viewmodels
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import com.automattic.simplenote.models.Note
+import com.automattic.simplenote.repositories.SimperiumCollaboratorsRepository
+import com.automattic.simplenote.repositories.TagsRepository
+import com.automattic.simplenote.usecases.GetTagsUseCase
+import com.automattic.simplenote.usecases.ValidateTagUseCase
+import com.simperium.client.Bucket
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.mockito.Mockito.mock
+import org.mockito.kotlin.any
+import org.mockito.kotlin.whenever
+
+class NoteEditorViewModelTest {
+    @get:Rule
+    val rule = InstantTaskExecutorRule()
+
+    private val tagsRepository: TagsRepository = mock(TagsRepository::class.java)
+    private val mockBucket: Bucket<*> = mock(Bucket::class.java)
+
+    private lateinit var viewModel: NoteEditorViewModel
+    private lateinit var note: Note
+
+    @Before
+    fun setup() {
+        whenever(tagsRepository.isTagValid(any())).thenReturn(true)
+        whenever(tagsRepository.isTagMissing(any())).thenReturn(true)
+
+        val collaboratorsRepository = SimperiumCollaboratorsRepository()
+        val getTagsUseCase = GetTagsUseCase(tagsRepository, collaboratorsRepository)
+        val validateTagUseCase = ValidateTagUseCase(tagsRepository, collaboratorsRepository)
+        viewModel = NoteEditorViewModel(getTagsUseCase, validateTagUseCase)
+
+        note = Note("key1")
+        note.content = "Hello World"
+        note.tags = listOf("tag1", "tag2", "name@test.com")
+        note.bucket = mockBucket
+    }
+
+    @Test
+    fun updateShouldUpdateUiState() {
+        viewModel.update(note)
+
+        assertEquals(listOf("tag1", "tag2"), viewModel.uiState.value?.tags)
+        assertEquals(listOf("tag1", "tag2", "name@test.com"), note.tags)
+    }
+
+    @Test
+    fun addTagShouldUpdateUiState() {
+        viewModel.addTag("tag3", note)
+
+        assertEquals(listOf("tag1", "tag2", "tag3"), viewModel.uiState.value?.tags)
+        assertEquals(listOf("tag1", "tag2", "name@test.com", "tag3"), note.tags)
+    }
+
+    @Test
+    fun addCollaboratorShouldNotUpdateUiStateAndNote() {
```

Comment:
> This is the single test assertion principle that I personally try to abide by.

Interesting, I haven't heard about it before. I will take a closer look into it to see how we can write more maintainable tests now that I think of. Fixed in e6f2a578815e0f72fb8e839a5044a9db23370c65.

Reply:
Oh well, this is not a principle per se, rather a rule of sorts. Doing a quick Google search will give you the: 

`"One assertion per test" is a wise rule to keep in mind, because it helps you have tests that fail for a specific reason, and drives you to focus on a specific behavior at a time. ... in his great writing “Test Desiderata”, this test is still “specific”, because, if it fails, “the cause of the failure should be obvious”.`

Thank you so much for even considering it! 🙇 

PS: To me, when I write tests, I write them in such a way that all test get to be closed to modification. So, if a logic gets updated, say something is added, I then want to add a new test, else, if say something gets removed, then I want to remove a test, but never modify it by removing or adding a line within an existing test, thus keeping all tests closed to modification, and only allowing deletion when necessary. Does this makes sense?